Target 8192 blocks instead of split to large grid for large reduction

zasdfgbnm opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36014 Refactor thread_reduce for better unrolling and vectorization in the future
* **#35997 Target 4096 blocks instead of split to large grid for large reduction**

When the number of blocks is large enough, we are already achieving
blalanced SM allocation. But we still should keep the number of inputs
per thread large, because thread reduce is cheap.

Benchmark for Half on V100:
https://github.com/zasdfgbnm/things/blob/master/2020Q2/reduction-benchmark.ipynb

On large tensor, it is: 1.37ms vs 1.25ms

Differential Revision: [D20927533](https://our.internmc.facebook.com/intern/diff/D20927533)

zasdfgbnm committed:
Target 4096 blocks instead of split to large grid for large reduction

When the number of blocks is large enough, we are already achieving
blalanced SM allocation. But we still should keep the number of inputs
per thread large, because thread reduce is cheap.

Benchmark for Half on V100:
https://github.com/zasdfgbnm/things/blob/master/2020Q2/reduction-benchmark.ipynb

On large tensor, it is: 1.37ms vs 1.25ms

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Target 4096 blocks instead of split to large grid for large reduction"


When the number of blocks is large enough, we are already achieving
blalanced SM allocation. But we still should keep the number of inputs
per thread large, because thread reduce is cheap.

Benchmark for Half on V100:
https://github.com/zasdfgbnm/things/blob/master/2020Q2/reduction-benchmark.ipynb

On large tensor, it is: 1.37ms vs 1.25ms

[ghstack-poisoned]

jjsjann123 commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -789,15 +789,18 @@ inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t id
-      config.ctas_per_output = 65535;
+    int ctas_per_output1 = div_up(config.values_per_thread(), 16);
+    int ctas_per_output2 = div_up(target_grid_size, grid);
+    config.ctas_per_output = std::min<int>(ctas_per_output1, ctas_per_output2);
```
Can we add a comment here explaining the motivation behind clamping down ctas_per_output.

zasdfgbnm replied:
fixed

jjsjann123 reviewed (approved):
LGTM. Please update the comment in code with your commit description.

zasdfgbnm committed:
Update on "Target 4096 blocks instead of split to large grid for large reduction"


When the number of blocks is large enough, we are already achieving
blalanced SM allocation. But we still should keep the number of inputs
per thread large, because thread reduce is cheap.

Benchmark for Half on V100:
https://github.com/zasdfgbnm/things/blob/master/2020Q2/reduction-benchmark.ipynb

On large tensor, it is: 1.37ms vs 1.25ms

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Target 4096 blocks instead of split to large grid for large reduction"


When the number of blocks is large enough, we are already achieving
blalanced SM allocation. But we still should keep the number of inputs
per thread large, because thread reduce is cheap.

Benchmark for Half on V100:
https://github.com/zasdfgbnm/things/blob/master/2020Q2/reduction-benchmark.ipynb

On large tensor, it is: 1.37ms vs 1.25ms

Differential Revision: [D20927533](https://our.internmc.facebook.com/intern/diff/D20927533)

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Target 4096 blocks instead of split to large grid for large reduction"


When the number of blocks is large enough, we are already achieving
blalanced SM allocation. But we still should keep the number of inputs
per thread large, because thread reduce is cheap.

Benchmark for Half on V100:
https://github.com/zasdfgbnm/things/blob/master/2020Q2/reduction-benchmark.ipynb

On large tensor, it is: 1.37ms vs 1.25ms

Differential Revision: [D20927533](https://our.internmc.facebook.com/intern/diff/D20927533)

[ghstack-poisoned]
